### ISSUE TITLE
Conditional import broken in Dart 2

### ISSUE THREAD

**Issue by daniel-v** [dart-lang/sdk#34262]
**Dart VM version**: 2.0.0 (Fri Aug 3 10:53:23 2018 +0200) on "windows_x64"

Conditional import/export worked in last version of Dart 1 and doesn't seem to work any longer.

```dart
export 'feature/myinterface.dart' if (somevar) 'impls/someimpl.dart';
```
Invoking it 
```bash
$> dart -Dsomevar=true bin\main.dart
```

Interestingly enough, I see it working with built-in libraries. `dart.library.io` evals while custom env values do not.

This prevents me from upgrading to Dart 2 in some of my projects.

**Comment by lrhn** [comment]
The version of conditional imports that made it into Dart 2 only allows the `dart.library.x` directives.
Allowing arbitrary directives intereres with staged modular compilation (the front-end compiler doesn't necessarily know what variables will be set to what). It means that you can only really use the conditional import/export feature for selecting code based on the backend (VM or browser) that the code is being compiled for.

**Comment by zoechi** [comment]
@lrhn I thought remembering such a discussion but couldn't find anything in the changelog.
Do you think it might be worth mentioning?
When the experimental feature was added it was mentioned https://github.com/dart-lang/sdk/blob/master/CHANGELOG.md#1150---2016-03-09

**Comment by natebosch** [comment]
Yeah this was discussed on https://github.com/dart-lang/sdk/issues/32960

tl;dr; config specific imports are only supported on `dart.library.*` conditions. I'm not sure if we have an issue tracking the language specification changes for this restriction.

**Comment by lrhn** [comment]
That specification just says that implementations may restrict which "environment variables" that can be used for switching. Our tools do so by only allowing `dart.library.x`.

**Comment by UNIcodeX** [comment]
Based on the above... I'm trying to build depending on the presence / absence of a dart file. The idea being that one branch has the file (pro version) and the other branch does not (free version), but I'm not doing something correctly. I have a file called isPro.dart in the same directory as another file which has the contents

```dart
import 'free.dart' if (isPro) 'pro.dart';
```

but that isn't working

Any ideas on what I can do?

If I test changing the condition to `(dart.library.io)` as in many examples, it imports the pro.dart contents, BUT, this isn't the condition I need checked. 

In essence, I'm trying to work within the new spec to emulate old functionality. Any ideas??

**Comment by natebosch** [comment]
> I'm trying to work within the new spec to emulate old functionality. Any ideas??

You will not be able to get that functionality from conditional imports.

Since you are already using source file changes as the signal for the change, I would recommend having aversion of an "alias" file on each branch. Use `import 'alias.dart';` as the import statement, then the file `alias.dart` will contain either `export 'free.dart';` or `export 'pro.dart';` depending on the branch.

**Comment by UNIcodeX** [comment]
ah! Thanks for that perspective. I was too close to the problem I suppose.

**Comment by liusong1111** [comment]
It's 2022, since flutter supports desktop and more, so that's unreasonable if this feature is still simply ignored

**Comment by nwparker** [comment]
> It's 2022, since flutter supports desktop and more, so that's unreasonable if this feature is still simply ignored

If you feel this way, why don't you open a feature request issue?

Otherwise it's not tracked and will probably never be done, right?

**Comment by banool** [comment]
@natebosch how do you do this? Something like this isn't allowed at the top level:
```
if (kIsWeb) {
   export 'my_module_web.dart';
} else {
   export 'my_module_other.dart';
}
```

**Comment by natebosch** [comment]
https://github.com/dart-lang/sdk/issues/34262#issuecomment-416778859

The valid conditions for imports are `dart.library.*`

**Comment by yancaidev** [comment]
**Maybe** we don't need conditional import any more due to [the-tree-shaking-mechanism-in-flutter](https://www.alibabacloud.com/blog/the-tree-shaking-mechanism-in-flutter_597737).  Tree-shaking will remove unused packages automatically.

Let me explain how I got the conclusion.

First, I made a flutter project named `flutter_tree_shaking` and two local packages within it.
<img width="1701" alt="compare_two_packages" src="https://user-images.githubusercontent.com/97107724/200235600-ff274c3c-26c8-4c92-86d6-7607017182f7.png">
 
Then I import these two packages to the project. 
<img width="1074" alt="import__packages" src="https://user-images.githubusercontent.com/97107724/200236063-fcd15587-e054-4388-928b-ac0a041c5625.png">

As the code above shows, what code will be called based on the environment variable `KEEP_FUNC1`.

Here's the important part. Run the project in profile mode without `KEEP_FUNC1` specified, execute `flutter build apk --analyze-size --target-platform=android-arm64` to generate `code-size-analysis` file , open devtools  in browser and import the code-size-analysis file into `App Size` tab. 

<img width="1858" alt="file_size_analysis" src="https://user-images.githubusercontent.com/97107724/200239235-535f684b-debf-4ea5-a228-31b2b34fe3a9.png">

See, only `package2` and the `addOne` method of `Calculator2` in `package2` exsist. Unused `page1` and its classes are removed.

 I'm not 100% sure of this. Looking forward a flutter master to ensure this.

**Comment by WorikQCI** [comment]
I have been hunting around for documentation for this technique:
```
import 'my_plugin_stub.dart'
    if (dart.library.io) 'my_plugin_io.dart'
    if (dart.library.html) 'my_plugin_web.dart';
```

That works well to do conditional compilation web V. other platforms.  (Early days for me yet, but we am planning to develop a web app in conjunction with mobile and desk top applications.  This looks like to be an important piece of the puzzle)

But I do not know what it is doing.  I can guess, but experience tells me that is a dangerous process.

What is this technique called?  Where is it documented?

**Comment by daniel-v** [comment]
🙏 @yancaidev  ! 

Very informative! Your approach has only 1 important difference: we are importing different symbols from the 2 packages.

Having 2 classes can easily be worked around just like you did but the return value type should be the same.

I'll play around with it for web too, I think tree shaking works the same.

**Comment by yancaidev** [comment]
[daniel-v](https://github.com/daniel-v)
This is an example to demonstrate how the tree shaking works. It's an another way to do conditional import. Returning same or different value types depends on the specific situation.
It's very common to import different types from different packages. If we need to use them as a same value type, defining an interface to bridge them together will be a good start.

**Comment by yancaidev** [comment]
[daniel-v](https://github.com/daniel-v)
This is an example to demonstrate how the tree shaking works. It's an another way to do conditional import. Returning same or different value types depends on the specific situation.
It's very common to import different types from different packages. If we need to use them as a same value type, defining an interface to bridge them together will be a good start.

**Comment by s4rath07** [comment]
Is there any way to conditionally import libraries based on web or android ??

**Comment by michasng** [comment]
Im my project, I created a file `platform_foo.dart` to execute platform specific code:

```dart
export 'foo.dart'
    if (dart.library.html) 'foo_web.dart'
    if (dart.library.io) 'foo_io.dart';
```
